src/mlpack/methods/ann/vanilla_rbm.hpp

```diff
@@ -0,0 +1,189 @@
+#include <mlpack/core/math/random.hpp>
+#include "layer/layer.hpp"
+#include "layer/base_layer.hpp"
+#include "visitor/weight_set_visitor.hpp"
+#include "visitor/forward_visitor.hpp"
+#include "visitor/weight_size_visitor.hpp"
+#include "activation_functions/softplus_function.hpp"
+#include "init_rules/gaussian_init.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InitializationRuleType,
+         typename VisibleLayerType,
+         typename HiddenLayerType>
+class VanillaRBM
+{
+ public:
+
+  using NetworkType =VanillaRBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>;
+
+  /* 
+   * Intalise all the parameters of the network
+   * using the intialise rule. 
+   *
+   * @tparam: IntialiserType rule to intialise the parameters of the netwokr
+   * @tparam: VisibleLayerType visible layer 
+   * @tparam: HiddenLyaerType hidden layer
+   */
+  VanillaRBM(InitializationRuleType initializeRule, VisibleLayerType visible, HiddenLayerType hidden);
+
+  // Reset the network
+  void Reset();
+
+  /* 
+   * Train the netwrok using the Opitimzer with given set of args.
+   * the optimiser sets the paratmeters of the network for providing
+   * most likely parameters given the inputs
+   * @param: predictors data points
+   * @param: optimizer Optimizer type
+   * @param: Args arguments for the optimizers
+   */
+  template<template <typename, typename...> class Optimizer, typename... OptimizerTypeArgs>
+  void Train(const arma::mat& predictors, Optimizer<NetworkType, OptimizerTypeArgs...> optimizer);
+ /* 
+  * This function calculates
+  * the free energy of the model
+  * @param: input data point 
+  */
+  double FreeEnergy(const arma::mat&& input);
+
+ /*
+  * This functions samples the hidden
+  * layer given the visible layer
+  *
+  * @param input visible layer input
+  * @param output the sampled hidden layer
+  */
+  void SampleHidden(arma::mat&& input, arma::mat&& output);
+
+  /*
+  * This functions samples the visble
+  * layer given the hidden layer
+  *
+  * @param input hidden layer
+  * @param output the sampled visible layer 
+  */
+  void SampleVisible(arma::mat&& input, arma::mat&& output);
+
+ /*
+  * This function does the k-step
+  * gibbs sampling.
+  *
+  * @param negative_sample: stores the negative sample
+  * @param num_step number of steps in gibbs sampling
+  * @param persistnce start chain at input or the previous negative_sample
+  */
+  void Gibbs(arma::mat&& input, 
+             arma::mat&& negative_sample, 
+             size_t num_steps = 1, 
+             bool persistence = false);
+
+
+  /*
+   * Calculates the gradients for the rbm network
+   *
+   * @param input the visible layer/data point
+   * @param neg_input stores the negative samples computed usign the gibbs 
+   * @param k number of steps for gibbs sampling
+   * @param persistence pcdk / not 
+   * @param output store the gradients
+   */
+  void Gradient(arma::mat input, 
```

Comment:
You can pass `k` and `persistence` at the constructor. If you do so, the method will look like other `Gradient()` methods.